### DISCUSSION TITLE
Jsonnet: Object or Array

### DISCUSSION THREAD

**Discussion by ghostsquad** [argoproj/argo-cd#15988] [Q&A]
this guide mentions ArgoCD will parse an object or an array. Does this mean that for an object, the keys can be arbitrary (like with Tanka)?

https://argo-cd.readthedocs.io/en/stable/user-guide/jsonnet/

I'd also love to see an example that shows this, and a link to the examples in the documentation. I think that would help other users understand the expectation when using Jsonnet. Thought I'd start a discussion first, then move to an issue or PR if other people are thinking the same thing. Thanks!

**Comment by ChristianCiach** [comment]
Looking at the code, it looks like a generated Object will be interpreted as if the output was a json file containing a single kubernetes manifest, while an Array will be interpreted as a set of json files, each containing a separate kubernetes manifest.

In other words, you can use a single jsonnet file to simulate a multi-document yaml file by generating a json array. 

I agree that the documentation leaves a lot to be desired. I also think that the documentation about jsonnet is just plain wrong in parts. For example, the documentation about the `directory` type applications state:

> Directory-type applications only work for plain manifest files. If Argo CD encounters Kustomize, Helm, or Jsonnet files when directory: is set, it will fail to render the manifests.

This seems wrong. `.jsonnet` files are treated the same as plain manifest files (with the exception that they are evaluated using Jsonnet, of course). In fact, there is no other way to use Jsonnet with ArgoCD. It can *only* be used with "directory" type applications.
